File: js/1987.js
const fs = require('fs');

const inputs = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const dy = [1, 0, -1, 0];
const dx = [0, 1, 0, -1];

const [r, c] = inputs[0].split(' ').map(Number);
const board = [];
const visited = Array(26).fill(false);

for (let i = 1; i <= r; i++) {
  board.push(inputs[i].split(''));
}

let answer = 1;
visited[board[0][0].charCodeAt() - 65] = true;

dfs(0, 0, 1);

console.log(answer);

function dfs(y, x, cnt) {
  if (cnt > answer) answer = cnt;
  for (let i = 0; i < 4; i++) {
    const ny = y + dy[i];
    const nx = x + dx[i];
    if (inRange(ny, nx) && !visited[board[ny][nx].charCodeAt() - 65]) {
      visited[board[ny][nx].charCodeAt() - 65] = true;
      dfs(ny, nx, cnt + 1);
      visited[board[ny][nx].charCodeAt() - 65] = false;
    }
  }
}

function inRange(y, x) {
  return 0 <= y && y < r && 0 <= x && x < c;
}

File: c++/1987.cpp
#include <iostream>
#include <string>

using namespace std;

int r, c;
int board[20][20];
bool visited[26];
int answer = 1;
int dy[4] = {1, 0, -1, 0};
int dx[4] = {0, 1, 0, -1};

bool inRange(int y, int x) { return 0 <= y && y < r && 0 <= x && x < c; }

void dfs(int y, int x, int cnt) {
  if (cnt > answer) answer = cnt;
  for (int i = 0; i < 4; i++) {
    int ny = y + dy[i];
    int nx = x + dx[i];
    if (inRange(ny, nx) && !visited[board[ny][nx] - 'A']) {
      visited[board[ny][nx] - 'A'] = true;
      dfs(ny, nx, cnt + 1);
      visited[board[ny][nx] - 'A'] = false;
    }
  }
}

int main() {
  // freopen("example.txt", "r", stdin);

  cin >> r >> c;
  for (int i = 0; i < r; i++) {
    string s;
    cin >> s;
    for (int j = 0; j < c; j++) {
      board[i][j] = s[j];
    }
  }

  visited[board[0][0] - 'A'] = true;
  dfs(0, 0, 1);

  cout << answer;

  return 0;
}
